cirq/google/serializable_gate_set.py

```diff
@@ -51,12 +49,35 @@ def __init__(self, gate_set_name: str,
                 forms of gates to GateOperations.
         """
         self.gate_set_name = gate_set_name
-        self.serializers = defaultdict(
-            list)  # type: Dict[Type, List[op_serializer.GateOpSerializer]]
+        self.serializers: Dict[Type, List[op_serializer.GateOpSerializer]] = {}
         for s in serializers:
-            self.serializers[s.gate_type].append(s)
+            self.serializers.setdefault(s.gate_type, []).append(s)
         self.deserializers = {d.serialized_gate_id: d for d in deserializers}
 
+    def with_gates(
+            self,
+            *,
+            name: Optional[str] = None,
+            serializers: Iterable[op_serializer.GateOpSerializer] = (),
+            deserializers: Iterable[op_deserializer.GateOpDeserializer] = (),
+    ) -> 'SerializableGateSet':
+        """Creates a new gateset with more (de)serializers.
+
+        Args:
+            name: Optional new name of the gateset. If not given, use the same
+                name as this gateset.
+            serializers: Serializers to add to those in this gateset.
+            deserializers: Deserializers to add to those in this gateset.
+        """
+        return SerializableGateSet(
+            name or self.gate_set_name,
+            serializers=[*self._all_serializers(), *serializers],
+            deserializers=[*self.deserializers.values(), *deserializers])
```

Comment:
Why do we need _all_serializers() but not deserializers?

Reply:
Serializers are stored as a map from gate_id to list of serializers, so we have to flatten those into one list. Deserializers are stored as a map from gate id to deserializer, so we can just use `.values()`. I will refactor to just inline the expression, since it's probably not worth having a whole method for the serializer flattening.